### ISSUE TITLE
Make compile fails on first invocation in develop branch

### ISSUE THREAD

**Issue by unbalancedparentheses** [lfe/lfe#188]
After the second invocation of make compile everything works fine. However on the first try I get an error. This only happens on the develop branch. At this moment the develop branch is the default git branch. So if a new user clones the repository he might find this issue.

I have Erlang 18.2.1-1 installed and rebar rebar 2.6.1 18 20151027_122802. I am using Arch Linux

``` bash
$ git clone https://github.com/rvirding/lfe.git
Cloning into 'lfe'...
remote: Counting objects: 3518, done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 3518 (delta 0), reused 0 (delta 0), pack-reused 3515
Receiving objects: 100% (3518/3518), 2.36 MiB | 1.05 MiB/s, done.
Resolving deltas: 100% (2153/2153), done.
Checking connectivity... done.

$ cd lfe

$ make compile
escript get_maps_opts.escript
if which rebar.cmd > /dev/null; \
then ERL_LIBS=.:$ERL_LIBS rebar.cmd compile; \
elif which rebar > /dev/null; \
then ERL_LIBS=.:$ERL_LIBS rebar compile; \
else \
make  erlc-lfec; \
fi
which: no rebar.cmd in (/home/unbalanced/.pyenv/shims:/home/unbalanced/.pyenv/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/unbalanced/.fzf/bin)
==> lfe (compile)
Compiled src/lfe_scan.xrl
Compiled src/lfe_lib.erl
Compiled src/lfe_env.erl
Compiled src/lfe_pmod.erl
Compiled src/lfe_macro_record.erl
Compiled src/lfe_trans.erl
src/lfe_ms.erl:302: Warning: function fetch_binding/2 is unused
Compiled src/lfe_ms.erl
Compiled src/lfe_io_pretty.erl
Compiled src/lfe_init.erl
src/lfe_codegen.erl:840: Warning: variable 'St' shadowed in 'fun'
src/lfe_codegen.erl:870: Warning: variable 'Map' is unused
src/lfe_codegen.erl:963: Warning: variable 'St' shadowed in 'fun'
src/lfe_codegen.erl:970: Warning: variable 'St' shadowed in 'fun'
src/lfe_codegen.erl:975: Warning: variable 'St' shadowed in 'fun'
Compiled src/lfe_codegen.erl
src/lfe_macro.erl:669: Warning: function trim_stacktrace/1 is unused
src/lfe_macro.erl:1045: Warning: variable 'V' shadowed in 'fun'
Compiled src/lfe_macro.erl
Compiled src/lfe_io.erl
Compiled src/lfe_io_write.erl
Compiled src/lfe_parse.erl
src/lfe_edlin_expand.erl:228: Warning: function over_non_symbol/3 is unused
Compiled src/lfe_edlin_expand.erl
src/lfe_comp.erl:671: Warning: function debug_print/3 is unused
Compiled src/lfe_comp.erl
Compiled src/lfe.erl
Compiled src/lfe_bits.erl
Compiled src/lfe_gen.erl
Compiled src/lfe_user_macros.erl
Compiled src/lfe_lint.erl
Compiled src/lfe_macro_include.erl
Compiled src/lfescript.erl
Compiled src/lfe_io_format.erl
Compiled src/lfe_scan.erl
src/lfe_eval.erl:128: Warning: variable 'M' shadowed in 'fun'
src/lfe_eval.erl:132: Warning: variable 'M' shadowed in 'fun'
src/lfe_eval.erl:462: Warning: function init_letrec_env/1 is unused
src/lfe_eval.erl:468: Warning: function extend_letrec_env/3 is unused
src/lfe_eval.erl:468: Warning: variable 'Fbs0' is unused
src/lfe_eval.erl:737: Warning: variable 'M' shadowed in 'fun'
src/lfe_eval.erl:741: Warning: variable 'M' shadowed in 'fun'
Compiled src/lfe_eval.erl
Compiled src/lfe_shell.erl
src/lfe_qlc.erl:282: Warning: erl_lint:modify_line/2: deprecated (will be removed in OTP 19); use erl_parse:map_anno/2 instead
src/lfe_qlc.erl:372: Warning: erl_parse:set_line/2: deprecated (will be removed in OTP 19); use erl_anno:set_line/2 instead
src/lfe_qlc.erl:439: Warning: erl_parse:set_line/2: deprecated (will be removed in OTP 19); use erl_anno:set_line/2 instead
src/lfe_qlc.erl:2470: Warning: erl_parse:set_line/2: deprecated (will be removed in OTP 19); use erl_anno:set_line/2 instead
src/lfe_qlc.erl:2473: Warning: erl_parse:set_line/2: deprecated (will be removed in OTP 19); use erl_anno:set_line/2 instead
src/lfe_qlc.erl:2476: Warning: erl_parse:get_attribute/2: deprecated (will be removed in OTP 19); use erl_anno:{column,line,location,text}/1 instead
src/lfe_qlc.erl:2651: Warning: erl_parse:set_line/2: deprecated (will be removed in OTP 19); use erl_anno:set_line/2 instead
src/lfe_qlc.erl:2680: Warning: erl_parse:get_attribute/2: deprecated (will be removed in OTP 19); use erl_anno:{column,line,location,text}/1 instead
src/lfe_qlc.erl:2682: Warning: erl_parse:set_line/2: deprecated (will be removed in OTP 19); use erl_anno:set_line/2 instead
src/lfe_qlc.erl:2694: Warning: erl_parse:get_attribute/2: deprecated (will be removed in OTP 19); use erl_anno:{column,line,location,text}/1 instead
src/lfe_qlc.erl:2696: Warning: erl_parse:set_line/2: deprecated (will be removed in OTP 19); use erl_anno:set_line/2 instead
src/lfe_qlc.erl:2711: Warning: erl_parse:set_line/2: deprecated (will be removed in OTP 19); use erl_anno:set_line/2 instead
src/lfe_qlc.erl:2715: Warning: erl_parse:get_attribute/2: deprecated (will be removed in OTP 19); use erl_anno:{column,line,location,text}/1 instead
src/lfe_qlc.erl:2722: Warning: erl_parse:get_attribute/2: deprecated (will be removed in OTP 19); use erl_anno:{column,line,location,text}/1 instead
src/lfe_qlc.erl:2726: Warning: erl_parse:get_attribute/2: deprecated (will be removed in OTP 19); use erl_anno:{column,line,location,text}/1 instead
Compiled src/lfe_qlc.erl
ERROR: 
*** MISSING LFE COMPILER ***
  You must do one of the following:
    a) Install LFE globally in your erl libs
    b) Add LFE as a dep for your project, eg:
       {lfe, "0.6.1",
        {git, "git://github.com/rvirding/lfe",
         {tag, "v0.6.1"}}}

Compiling src/cl.lfe failed:
ERROR: compile failed while processing /home/unbalanced/projects/lfe: rebar_abort
Makefile:78: recipe for target 'compile' failed
make: *** [compile] Error 1

$ make compile
if which rebar.cmd > /dev/null; \
then ERL_LIBS=.:$ERL_LIBS rebar.cmd compile; \
elif which rebar > /dev/null; \
then ERL_LIBS=.:$ERL_LIBS rebar compile; \
else \
make  erlc-lfec; \
fi
which: no rebar.cmd in (/home/unbalanced/.pyenv/shims:/home/unbalanced/.pyenv/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/unbalanced/.fzf/bin)
==> lfe (compile)
ecp: #(cl ())
Compiled src/cl.lfe
Compiling c_src/lfeexec.c

$ ./bin/lfe
Erlang/OTP 18 [erts-7.2.1] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false]

LFE Shell V7.2.1 (abort with ^G)
> (+ 1 1)
2
>
```

**Comment by yurrriq** [comment]
I'm having a similar issue on Debian and v0.10.1 now.

**Comment by oubiwann** [comment]
I would suspect that the cause here is the ordering of files to compile.
Basically, we want to have all .erl files compiled first -- then the cl.lfe
(and any other .lfe files we add). If this guess is correct, then the order
of files is not deterministic across OSes ...

I guess I should clarify: the make targets are set up so that all .erl
files are compiled first, and then the .lfe files are compiled. We have no
such guarantees when rebar is used. The only way to fix that is to put all
the files that need to be compiled first in the rebar.config file, and then
put the .lfe files after those. Maybe there's a better way to do this with
rebar?

**Comment by rvirding** [comment]
The problem in this case is an issue with rebar2 which KNOWS about .lfe files and how to compile them. Unfortunately it does this by calling lfec. So here while all the LFE files have been compiled the lfec, and lfe, programs have not been installed and so rebar2 cannot find it. Hence the error.

As far as I know there is no way around this.

With rebar3 we need to write our own handling so we should be able to handle it.

**Comment by mqsoh** [comment]
Why does the Makefile need rebar? It falls back on erlc anyway, which everyone has.

Is the build better with rebar?

**Comment by rvirding** [comment]
I don't know as I use erlc-lfec option here. Others added the rebar interface. At this level you probably don't need it as it has no dependencies and yo would generally not create a release with just lfe.

Would anyone complain a lot if I removed it and used erlc directly?

**Comment by yurrriq** [comment]
I think rebar3 can offer finer control, especially wrt the `lfe-compile` plugin, for those who want it, so unless its inclusion is causing issues, I'd suggest keeping it as an option.

**Comment by rvirding** [comment]
At this level I wonder if `rebar` can give us very much extra? We we are basically just compiling the core LFE files which are written in erlang and some (one) .lfe file. We have no deps, and here we never will, LFE is the dep. One reason to keep the `rebar` option is if it makes it easier to compile on non-Unix systems (Windows).

Perhaps not make it the default if it exists if we want to keep it.

**Comment by oubiwann** [comment]
Yeah, I've been wondering about the same thing. We still need a `rebar.config` since we want projects that use rebar (or, preferably rebar3) to be able to build LFE correctly. 

However, even if we remove rebar usage from our Makefile, we'll still run into problems of other projects who have an LFE dependency and are using rebar to build.

My vote is 1) to remove rebar from our Makefile, and 2) get this fixed for any projects that _do_ use rebar and have LFE as a dep.

I'm working on this right now, btw :-)

**Comment by oubiwann** [comment]
Okay, part of the error has been identified in issue #189: the `is_map` guard in `src/cl.lfe` is going to break compiles on pre-17 Erlang systems.

However, since the error reported in this ticket is for an Erlang 18 system, I'm somewhat at a loss ... the `is_map` guard shouldn't be a problem.

Also, `lfe_comp.erl` is getting compiled fairly soon (certainly before the error above is thrown), so that should also not be a problem... yet the first error shown is indicative of rebar not actually being able to _find_ `lfe_comp.beam`. I _believe_ that the `Makefile` is properly setting the `ERL_LIBS` env variable (I tested this on a fresh Ubuntu install with R16B03) for rebar, so the mystery is a bit compounded.

**Comment by oubiwann** [comment]
Oh, I mis-read the output above -- the guard is not a problem.

I'll download an Arch Docker image and give the build a try there ...

**Comment by oubiwann** [comment]
I just built and pushed the Arch Docker image for LFE with no issues:
- https://hub.docker.com/r/lfex/arch/

That being said, the `Dockerfile` builds LFE before `rebar` is installed.

I just did a `make clean` in the LFE directory on Arch. With `rebar` installed, I then did a make, and it built using `rebar` without issue. Here are the versions of everything I have on Arch:
- erts 7.1
- rebar 2.6.1 R15B03 20150928_141254 git 2.6.1
- LFE v0.11.0-dev

If you can provide any more info about your setup that could help me duplicate your issue, that would be super-helpful ...

**Comment by mqsoh** [comment]
You can replicate OP's issue like this:

```
docker run -it --rm erlang:18 bash -c 'git clone https://github.com/rvirding/lfe.git; cd lfe; make'
```

The official Erlang image has rebar and rebar3. This issue is definitely rebar 2 related. (I think that's been established, though.)

```
docker run -it --rm erlang:18 bash -c 'rm $(which rebar); git clone https://github.com/rvirding/lfe.git; cd lfe; make'
```

**Comment by oubiwann** [comment]
Thanks, @mqsoh ... pulling the Erlang docker image now to give that a try ...

**Comment by oubiwann** [comment]
This should be fixed now -- please test and let us know!

(Note, Docker images haven't been updated yet -- on the docker for tonight.)

**Comment by oubiwann** [comment]
Docker images are pushed.

I don't see these errors anymore, due to the removal of rebar from the Makefile.

Could folks test and let us know? (either against the Docker images or against the latest -- and greatest!!! -- LFE release)

**Comment by yurrriq** [comment]
I'm nearly certain this is solved now, though I've not tested the Docker images.

**Comment by yurrriq** [comment]
I'm even more convinced this is resolved now...

**Comment by oubiwann** [comment]
@rvirding This was fixed in #198, but I don't have perms to close -- can you close this ticket?

**Comment by oubiwann** [comment]
Updated the status of this ticket ... looking at the release history to see if it has been.

**Comment by oubiwann** [comment]
Yup, released in 1.0.
